### PR TITLE
chore(asm): keep waf context during request

### PR DESCRIPTION
**Pull request by christophe-papazian**
To avoid performance issues, this PR defines a way to keep the same WAF context for all calls to the WAF during a single request.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: ddtrace/appsec/ddwaf/__init__.py [hunk 1]
```diff
@@ -126,33 +139,34 @@ def run(
...
-                LOGGER.error("DDWaf failure to create the context")
+            if self._ctx == 0:
+                LOGGER.warning("DDWaf failsafe to create the context")
+                self._ctx = ddwaf_context_init(self._handle)
+
+            if self._ctx == 0:
+                LOGGER.error("DDWaf failure: no context created")
                 return DDWaf_result(None, [], 0, (time.time() - start) * 1e6)
 
+            result = ddwaf_result()
```
Comment (gnufede) [inline ddtrace/appsec/ddwaf/__init__.py:150]: This block was previously inside a try block, are we sure about removing that?

Comment (christophe-papazian) [reply ddtrace/appsec/ddwaf/__init__.py:150]: Yes, the try/finally was only present to free the context. We don't want that anymore.

**Review by gnufede — APPROVED** [review]

File: ddtrace/appsec/ddwaf/__init__.py [hunk 2]
```diff
@@ -126,33 +139,34 @@ def run(
             # type: (...) -> DDWaf_result
             start = time.time()
 
-            ctx = ddwaf_context_init(self._handle)
-            if ctx == 0:
-                LOGGER.error("DDWaf failure to create the context")
+            if self._ctx == 0:
+                LOGGER.warning("DDWaf failsafe to create the context")
```
Comment (avara1986) [inline ddtrace/appsec/ddwaf/__init__.py:143]: is not this validation redundant with `_at_request_start`?

**Review by juanjux — APPROVED** [review]

File: ddtrace/appsec/ddwaf/__init__.py (hunk 2 above)
Comment (christophe-papazian) [reply ddtrace/appsec/ddwaf/__init__.py:143]: It's a failsafe. Yes it's redondant but in the case that span_start was not called for any reason before the call to the WAF, it allows to process as usual without errors.

**Pull request merged**